test/integration/scheduler/scheduler_test.go

```diff
@@ -105,7 +105,62 @@ var _ = ginkgo.Describe("Scheduler", func() {
-		gomega.Expect(createdProdJob2.Spec.Template.Spec.NodeSelector[labelKey]).Should(gomega.Equal(onDemandFlavor))
+		gomega.Expect(createdProdJob2.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor))
+	})
+
+	ginkgo.It("Should schedule jobs on tolerated flavors", func() {
+		ginkgo.By("creating capacity and queue")
+		capacity := testing.MakeCapacity("tainted-capacity").
+			Resource(testing.MakeResource(corev1.ResourceCPU).
+				Flavor(testing.MakeFlavor(spotFlavor, "5").
+					Taint(corev1.Taint{
+						Key:    instanceKey,
+						Value:  spotFlavor,
+						Effect: corev1.TaintEffectNoSchedule,
+					}).
+					Label("cloud.provider.com/instance", spotFlavor).Obj()).
+				Flavor(testing.MakeFlavor(onDemandFlavor, "5").
+					Label("cloud.provider.com/instance", onDemandFlavor).Obj()).
+				Obj()).
+			Obj()
+		gomega.Expect(k8sClient.Create(ctx, capacity)).Should(gomega.Succeed())
+		queue := testing.MakeQueue("tainted-queue", namespace).Capacity(capacity.Name).Obj()
+		gomega.Expect(k8sClient.Create(ctx, queue)).Should(gomega.Succeed())
+
+		ginkgo.By("checking a job without toleration starts on the non-tainted flavor")
+		job1 := testing.MakeJob("on-demand-job1", namespace).Queue(queue.Name).AddResource(corev1.ResourceCPU, "5").Obj()
+		gomega.Expect(k8sClient.Create(ctx, job1)).Should(gomega.Succeed())
+		createdJob1 := &batchv1.Job{}
+		gomega.Eventually(func() bool {
+			lookupKey := types.NamespacedName{Name: job1.Name, Namespace: job1.Namespace}
+			return k8sClient.Get(ctx, lookupKey, createdJob1) == nil && !*createdJob1.Spec.Suspend
+		}, timeout, interval).Should(gomega.BeTrue())
+		gomega.Expect(createdJob1.Spec.Template.Spec.NodeSelector[instanceKey]).Should(gomega.Equal(onDemandFlavor))
+
+		ginkgo.By("checking a second job without toleration doesn't start")
+		job2 := testing.MakeJob("on-demand-job2", namespace).Queue(queue.Name).AddResource(corev1.ResourceCPU, "5").Obj()
+		gomega.Expect(k8sClient.Create(ctx, job2)).Should(gomega.Succeed())
+		createdJob2 := &batchv1.Job{}
+		gomega.Consistently(func() bool {
+			lookupKey := types.NamespacedName{Name: job2.Name, Namespace: job2.Namespace}
+			return k8sClient.Get(ctx, lookupKey, createdJob2) == nil && *createdJob2.Spec.Suspend
+		}, consistentDuration, interval).Should(gomega.BeTrue())
+
+		ginkgo.By("checking a third job with toleration starts")
+		job3 := testing.MakeJob("spot-job3", namespace).Queue(queue.Name).
+			Toleration(corev1.Toleration{
+				Key:      instanceKey,
+				Operator: corev1.TolerationOpEqual,
+				Value:    spotFlavor,
+				Effect:   corev1.TaintEffectNoSchedule,
+			}).
+			AddResource(corev1.ResourceCPU, "5").Obj()
+		gomega.Expect(k8sClient.Create(ctx, job3)).Should(gomega.Succeed())
+		createdJob3 := &batchv1.Job{}
+		gomega.Eventually(func() bool {
+			lookupKey := types.NamespacedName{Name: job3.Name, Namespace: job3.Namespace}
+			return k8sClient.Get(ctx, lookupKey, createdJob3) == nil && *createdJob3.Spec.Suspend
```

Comment:
I find a strange thing here. According to my understanding，the `createdJob3` here should be `unsuspend` because of the `toleration` for `spot resource`. But in fact, the `createdJob3` can't be scheduled successfully. I've tested it many times locally and it works like this. After that, I find it is because we just pick the head of each queue.
https://github.com/kubernetes-sigs/kueue/blob/da1c239b0e4f47fdd797fa69ae626dc7dd06b129/pkg/queue/manager.go#L227-L239

Then job2 will always block job3. I don't mean to say that this is a bug. But we need to do something to prevent this. There are two solutions I might think of 1. Add a `schedulerAttemps` to record the failure counts. 2. Add a backoff mechanism mentioned here #8  Just personal opinion here.

Reply:
Indeed this is supposed to test that the job starts as the "By" log indicates. I fixed that and now job3 doesn't schedule unless I comment out job2.

We do need a fix for requeueing that prevents this starvation problem, which could happen not only because of taints, but also if job2 resource requests don't fit the capacity but job3's resource requests do. 

Before discussing solutions (kube-scheduler provides us with example solutions including backoff), we should discuss requirements and what semantics we would like to have. For example, to maintain FIFO scheduling, we need to make sure that the mechanism we come up with doesn't starve older jobs. Lets start discussing that at https://github.com/kubernetes-sigs/kueue/issues/8